src/backend/distributed/worker/worker_data_fetch_protocol.c

```diff
@@ -1031,45 +1033,60 @@ ParseTreeNode(const char *ddlCommand)
 Datum
 worker_append_table_to_shard(PG_FUNCTION_ARGS)
 {
-	text *shardNameText = PG_GETARG_TEXT_P(0);
-	text *remoteTableNameText = PG_GETARG_TEXT_P(1);
+	text *shardQualifiedNameText = PG_GETARG_TEXT_P(0);
+	text *remoteQualifiedNameText = PG_GETARG_TEXT_P(1);
 	text *nodeNameText = PG_GETARG_TEXT_P(2);
 	uint32 nodePort = PG_GETARG_UINT32(3);
 
-	char *shardName = text_to_cstring(shardNameText);
-	char *remoteTableName = text_to_cstring(remoteTableNameText);
+	List *shardQualifiedNameList = textToQualifiedNameList(shardQualifiedNameText);
+	List *remoteQualifiedNameList = textToQualifiedNameList(remoteQualifiedNameText);
 	char *nodeName = text_to_cstring(nodeNameText);
 
+	char *shardTableName = NULL;
+	char *shardSchemaName = NULL;
+	char *shardQuotedQualifiedName = NULL;
+	char *remoteSchemaName = NULL;
+	char *remoteTableName = NULL;
+	char *remoteQuotedQualifiedName = NULL;
+
 	StringInfo shardNameString = NULL;
 	StringInfo localFilePath = NULL;
 	StringInfo remoteCopyCommand = NULL;
 	CopyStmt *localCopyCommand = NULL;
 	RangeVar *localTable = NULL;
 	uint64 shardId = INVALID_SHARD_ID;
 	bool received = false;
-	char *quotedTableName = NULL;
 	StringInfo queryString = NULL;
-	const char *schemaName = NULL;
 
-	/* copy remote table's data to this node */
-	shardNameString = makeStringInfo();
-	appendStringInfoString(shardNameString, shardName);
+	/* We extract schema names and table names from qualified names */
+	DeconstructQualifiedName(shardQualifiedNameList,
+							 &shardSchemaName,
+							 &shardTableName);
+
+	DeconstructQualifiedName(remoteQualifiedNameList,
+							 &remoteSchemaName,
+							 &remoteTableName);
 
 	/*
 	 * We lock on the shardId, but do not unlock. When the function returns, and
 	 * the transaction for this function commits, this lock will automatically
 	 * be released. This ensures appends to a shard happen in a serial manner.
 	 */
+	shardNameString = makeStringInfo();
```

Comment:
We create shardNameString only to be able to extract shardId and thus put a lock on shardId. ExtractShardId function only works on StringInfo type. Therefore creating shardNameString is part of locking procedure.